src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,166 @@
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
+    private static final class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T item) {
+            this.item = item;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = newNode;
+            tail = newNode;
+            size++;
+        } else {
+            addLast(newNode);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " must be between 0 and size "
+                    + size + " inclusively");
+        }
+        Node<T> next = getNode(index);
+        Node<T> newNode = new Node<>(value);
+        if (next.prev == null) {
+            next.prev = newNode;
+            head = newNode;
+            newNode.next = next;
+        } else {
+            next.prev.next = newNode;
+            newNode.prev = next.prev;
+            newNode.next = next;
+            next.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T node : list) {
+                add(node);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNode(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> nodeByIndex = getNode(index);
+        T oldValue = nodeByIndex.item;
+        nodeByIndex.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNode(index);
+        return unlink(node);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        int index = 0;
+        while (node != null) {
+            if (Objects.equals(node.item, object)) {
+                unlink(getNode(index));
+                return true;
+            }
+            node = node.next;
+            index++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(index + " index too big for the size " + size);
```

Comment:
I think that here it is necessary to change the messages,  if the index is less than zero?

Reply:
thanks for review)